django/test/testcases.py

```diff
@@ -583,7 +583,11 @@ def _assert_contains(self, response, text, status_code, msg_prefix, html):
         content_repr = safe_repr(content)
         if not isinstance(text, bytes) or html:
             text = str(text)
-            content = content.decode(response.charset)
+            content = (
+                content.decode(response.charset)
+                if response.streaming
+                else response.text
+            )
```

Comment:
I think as this handles streaming response, I am tempted to leave this one how it was. 

But the following I think we can update:
```diff
--- a/django/test/client.py
+++ b/django/test/client.py
@@ -947,9 +947,7 @@ class ClientMixin:
                     'Content-Type header is "%s", not "application/json"'
                     % response.get("Content-Type")
                 )
-            response._json = json.loads(
-                response.content.decode(response.charset), **extra
-            )
+            response._json = json.loads(response.text, **extra)
         return response._json
 
     def _follow_redirect(
diff --git a/tests/admin_views/tests.py b/tests/admin_views/tests.py
index fc1bb86d85..e95c2c0804 100644
--- a/tests/admin_views/tests.py
+++ b/tests/admin_views/tests.py
@@ -298,7 +298,7 @@ class AdminViewBasicTestCase(TestCase):
             response.content.index(text2.encode()),
             (failing_msg or "")
             + "\nResponse:\n"
-            + response.content.decode(response.charset),
+            + response.text,
         )
```

Reply:
Hi! @sarahboyce  :) Thank you as always for your reviews.

Regarding the part you reviewed, I think that in cases where it's not `response.streaming`, we can use `response.text` as shown in the link below. 

https://github.com/django/django/blob/db1419e58bec9c6ce2b006e799fbb315c5697568/django/test/testcases.py#L579-L582

Reply:
What would that look like? 
I still think we might want to leave this one :thinking: I'm not convinced it will improve it much

Reply:
I simply thought that the method related to `decode` could be replaced with `text`.

I agree that there's uncertainty about whether it will significantly improve things. If there's a lack of confidence, it might be necessary to consider removing it.

Thank you for taking the time to review this! 😊